zerver/actions/reminders.py

```diff
@@ -0,0 +1,122 @@
+            topic=message.topic_name(),
+        )
+    elif message.recipient.type == message.recipient.PERSONAL:
+        narrow_link = personal_narrow_url(
+            realm=message.sender.realm,
+            sender=message.sender,
+        )
+    else:
+        narrow_link = huddle_narrow_url(
+            user=current_user,
+            display_recipient=get_display_recipient(message.recipient),
+        )
+
+    return f"{narrow_link}/near/{message.id}"
+
+
+def get_formatted_content_from_message(message: Message, current_user: UserProfile) -> str:
+    if message.is_stream_message():
+        stream_id = message.recipient.type_id
+        stream = Stream.objects.get(id=stream_id)
+        narrow_link = topic_narrow_url(
+            realm=message.sender.realm,
+            stream=Stream.objects.get(id=stream_id),
+            topic=message.topic_name(),
+        )
+        content = "You requested a reminder for the following message sent to "
+        content += f"[{stream.name} > {message.topic_name()}]({narrow_link})."
+    else:
+        content = "You requested a reminder for the following private message."
+    content += "\n\n"
+    content += _("@_**{sender_full_name}|{sender_id}** [said]({conversation_url}):").format(
+        sender_full_name=message.sender.full_name,
+        sender_id=message.sender.id,
+        conversation_url=by_conversation_and_time_url(message, current_user),
+    )
+    content += "\n"
+    fence = get_unused_fence(content)
+    quoted_message = f"{fence}quote\n{message.content}\n{fence}"
+    content += quoted_message
+    return content
+
+
+def check_remind_me_message(
+    current_user: UserProfile,
+    client: Client,
+    message_id: int,
+    deliver_at: datetime.datetime,
+) -> int:
+    try:
+        message, _ = access_message(current_user, message_id)
+    except JsonableError as error:
+        internal_send_private_message(
+            get_system_bot(settings.NOTIFICATION_BOT, current_user.realm.id),
+            current_user,
+            f"You asked for a reminder about a message with ID {message_id}, but either the message has been deleted or you no longer have access to the message.",
+        )
+        raise error
+
+    notification_bot = get_system_bot(settings.NOTIFICATION_BOT, current_user.id)
+    addressee = Addressee.legacy_build(
+        # send_scheduled_message sends REMIND message from NOTIFICATION_BOT
+        # regardless of the value passed here.
+        notification_bot,
+        "private",
+        [current_user.id],
+        None,
+        current_user.realm,
+    )
+    send_request = check_message(
+        notification_bot,
+        client,
+        addressee,
+        get_formatted_content_from_message(message, current_user),
+        realm=current_user.realm,
+        forwarder_user_profile=current_user,
+    )
+    send_request.deliver_at = deliver_at
+
+    return do_schedule_messages([send_request], notification_bot, ScheduledMessage.REMIND)[0]
```

Comment:
I think ideally we'd have the rewriting to `notification_bot` as the actual sender happen inside `send_scheduled_message` -- we need to keep track of the user who requested the reminder in order to do permissions/management of the `ScheduledMessage` objects.

Reply:
Ok, done.